### ISSUE TITLE
Can't add setting when config has been accessed

### ISSUE THREAD

**Issue by waiting-for-dev** [dry-rb/dry-configurable#70]
Way to reproduce:

```ruby
class A
  extend Dry::Configurable

  setting :a
end

A.config

class A
  setting :b
end

# => Dry::Configurable::AlreadyDefinedConfig (Cannot add setting +b+, A is already configured)
```

If I omit the `A.config` in-between, there is no problem in adding setting `:b`.

I don't think it is desirable. In my scenario it happens because I'm developing a library with a bunch of extensions. Some of these extensions reopen a class to add a new `setting`. While surely it would be no problem in production (users will load all needed extensions at once, before any call to `#config`) it is a problem for my test suite, as I pick extensions depending on the test file and some tests access `#config` to make an assertion. I know I could remove the constant after each test, but anyway I don't think it is the desired behavior.

**Comment by flash-gordon** [comment]
Have you checked https://dry-rb.org/gems/dry-configurable/testing/ ?

**Comment by waiting-for-dev** [comment]
I hadn't, and the funny thing is that it was a [PR](https://github.com/dry-rb/dry-configurable/pull/24) of mine which introduced it :D

However, it doesn't change anything for this scenario:

```ruby
class A
  extend Dry::Configurable

  setting :a
end

A.config
require 'dry/configurable/test_interface'
A.enable_test_interface
A.reset_config

class A
  setting :b
end

# => Dry::Configurable::AlreadyDefinedConfig (Cannot add setting +b+, A is already configured)
```

**Comment by flash-gordon** [comment]
Hm, we could fix it by resetting the `@config_defined` variable. Allowing adding settings dynamically though possible (as far as I can see) is 1) a breaking change 2) somewhat more involved, I wouldn't do it without a solid reason, it never was an issue beyond tests (again, AFAIK).

**Comment by waiting-for-dev** [comment]
> somewhat more involved, I wouldn't do it without a solid reason

Yeah, I also felt uneasy with that. Finally, I have found a better way which doesn't need to reopen classes. So, at least for now, I'm going to close it.

Thank your for your help.

**Comment by solnic** [comment]
Re-opening this it's one of the caveats I want to address before 1.0.0. ie dry-system could benefit from this and possibly many more gems.

**Comment by waiting-for-dev** [comment]
@solnic a caveat you could encounter with this is that settings won't be added to subclasses when they haven been loaded before the new setting is added:

```ruby
class A
  extend Dry::Configurable
end

class B < A
end

A.setting :a

A.settings # => #<Set: {:a}>
B.settings # => #<Set: {:}>

class C < A
end

C.settings  # => #<Set: {:a}>

**Comment by dorner** [comment]
One use case I have for this is a setting whose default value should be set to the value of a different setting:
```ruby
  class A extend Dry::Configurable
    setting :foo, 5
    setting :bar, self.config.foo
  end
```

**Comment by solnic** [comment]
I fixed it for real in #77

**Comment by solnic** [comment]
oops, that's a different use case, re-opening 🙄

**Comment by solnic** [comment]
This was fixed in `0.11.0`

**Comment by solnic** [comment]
@dorner your usecase works now too with `0.11.0`
